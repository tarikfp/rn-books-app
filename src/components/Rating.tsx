import React from 'react';
import { StyleSheet } from 'react-native';
import Icon from 'react-native-vector-icons/FontAwesome';
import { useAppTheme } from '~theme/core';
import { Box } from './Box';

const ICON_SIZE = 20;

interface RatingProps {
  rating: number;
  iconSize?: number;
  testID?: string;
}

export const StarRating: React.FC<RatingProps> = ({
  rating,
  testID,
  iconSize = ICON_SIZE,
}) => {
  const { colors } = useAppTheme();
  const renderStars = () => {
    const stars = [];
    const fullStars = Math.floor(rating);
    const halfStar = rating - fullStars >= 0.5;

    for (let i = 1; i <= 5; i++) {
      let iconName = 'star';
      let iconColor = colors.bodyTextDisabled;

      if (i <= fullStars) {
        iconName = 'star';
        iconColor = colors.gold;
      } else if (i === fullStars + 1 && halfStar) {
        iconName = 'star-half-o';
        iconColor = colors.gold;
      }

      stars.push(
        <Icon
          testID={(testID ?? '') + i}
          key={i}
          name={iconName}
          size={iconSize}
          color={iconColor}
          style={styles.star}
        />,
      );
    }

    return stars;
  };

  return (
    <Box flexDirection={'row'} alignItems={'center'}>
      {renderStars()}
    </Box>
  );
};

const styles = StyleSheet.create({
  star: {
    marginHorizontal: 2,
  },
});
