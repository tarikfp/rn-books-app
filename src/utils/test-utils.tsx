import { ThemeProvider } from '@shopify/restyle';
import { render, RenderOptions } from '@testing-library/react-native';
import renderer from 'react-test-renderer';
import { theme } from '~theme/core';
import { ThemePreferenceContext } from '~theme/useThemePreference';

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <ThemePreferenceContext.Provider
      value={{
        isDarkTheme: false,
        toggleTheme: () => {},
      }}
    >
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
    </ThemePreferenceContext.Provider>
  );
};

const withAllTheProviders = (ui: React.ReactNode) => (
  <AllTheProviders children={ui} />
);

const customRender = (ui: React.ReactElement, options?: RenderOptions) =>
  render(ui, { wrapper: AllTheProviders, ...options });

const customRenderer = (ui: React.ReactElement) =>
  renderer.create(withAllTheProviders(ui));

// re-export everything
export * from '@testing-library/react-native';
// override render method
export { customRender as render };
// override renderer method
export { customRenderer as createRenderer };
