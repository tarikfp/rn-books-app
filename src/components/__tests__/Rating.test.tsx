import React from 'react';
import { StarRating } from '~components/Rating';
import { render } from '~utils/test-utils';

describe('StarRating Component', () => {
  it('matches snapshot', () => {
    const { toJSON } = render(<StarRating rating={3.5} />);
    expect(toJSON()).toMatchSnapshot();
  });

  it('renders the correct number of stars', () => {
    const { getByTestId } = render(<StarRating testID="star" rating={4} />);

    for (const star of ['star1', 'star2', 'star3', 'star4', 'star5']) {
      expect(getByTestId(star)).toBeTruthy();
    }
  });
});
