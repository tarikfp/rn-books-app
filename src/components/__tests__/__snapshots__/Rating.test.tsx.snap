// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StarRating Component matches snapshot 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "flexDirection": "row",
      },
    ]
  }
>
  <Text
    allowFontScaling={false}
    selectable={false}
    style={
      [
        {
          "color": "#FFD93D",
          "fontSize": 20,
        },
        {
          "marginHorizontal": 2,
        },
        {
          "fontFamily": "FontAwesome",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        {},
      ]
    }
    testID="1"
  >
    
  </Text>
  <Text
    allowFontScaling={false}
    selectable={false}
    style={
      [
        {
          "color": "#FFD93D",
          "fontSize": 20,
        },
        {
          "marginHorizontal": 2,
        },
        {
          "fontFamily": "FontAwesome",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        {},
      ]
    }
    testID="2"
  >
    
  </Text>
  <Text
    allowFontScaling={false}
    selectable={false}
    style={
      [
        {
          "color": "#FFD93D",
          "fontSize": 20,
        },
        {
          "marginHorizontal": 2,
        },
        {
          "fontFamily": "FontAwesome",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        {},
      ]
    }
    testID="3"
  >
    
  </Text>
  <Text
    allowFontScaling={false}
    selectable={false}
    style={
      [
        {
          "color": "#FFD93D",
          "fontSize": 20,
        },
        {
          "marginHorizontal": 2,
        },
        {
          "fontFamily": "FontAwesome",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        {},
      ]
    }
    testID="4"
  >
    
  </Text>
  <Text
    allowFontScaling={false}
    selectable={false}
    style={
      [
        {
          "color": "#B4B6BB",
          "fontSize": 20,
        },
        {
          "marginHorizontal": 2,
        },
        {
          "fontFamily": "FontAwesome",
          "fontStyle": "normal",
          "fontWeight": "normal",
        },
        {},
      ]
    }
    testID="5"
  >
    
  </Text>
</View>
`;
